File: programmers/양궁대회/Programmers92342.java
package com.example;

import java.util.Arrays;

public class Programmers92342_2 {

    private int maxScoreGap;
    private int[] lionMaxArr;

    public int[] solution(int n, int[] info) {
        maxScoreGap = -1;
        lionMaxArr = new int[info.length];
        recursion(n, info, new int[info.length], 0);
        return maxScoreGap == -1 ? new int[]{-1} : lionMaxArr;
    }

    private void recursion(int n, int[] apeach, int[] lion, int idx) {
        if (apeach.length == idx) {
            lion[idx - 1] = n;
            compare(apeach, lion);
            return;
        }

        if (apeach[idx] < n) {
            lion[idx] = apeach[idx] + 1;
            recursion(n - lion[idx], apeach, lion, idx + 1);
            lion[idx] = 0;
        }
        recursion(n, apeach, lion, idx + 1);
    }

    private void compare(int[] apeach, int[] lion) {
        int apeachScore = 0, lionScore = 0;
        for (int i = 0; i < apeach.length; i++) {
            if (apeach[i] == 0 && lion[i] == 0) continue;
            if (apeach[i] >= lion[i]) apeachScore += 10 - i;
            else lionScore += 10 - i;
        }

        if (apeachScore >= lionScore) return;

        int scoreGap = lionScore - apeachScore;
        if (maxScoreGap < scoreGap) {
            maxScoreGap = scoreGap;
            lionMaxArr = Arrays.copyOf(lion, lion.length);
        } else if (maxScoreGap == scoreGap) {
            if (canUpdateLionNumMaxArr(lion)) lionMaxArr = Arrays.copyOf(lion, lion.length);
        }
    }

    private boolean canUpdateLionNumMaxArr(int[] lion) {
        for (int i = lion.length - 1; i > -1; i--) {
            if (lionMaxArr[i] > lion[i]) return false;
            if (lionMaxArr[i] < lion[i]) return true;
        }
        return false;
    }
}

File: programmers/양궁대회/Programmers92342.kt
package com.example

import kotlin.math.min

class Programmers92342 {
    companion object {
        var maxDiff = 0
        var maxDiffArray = intArrayOf(-1)
    }

    fun solution(n: Int, info: IntArray): IntArray {
        recursion(n, info, IntArray(info.size), 0)
        return maxDiffArray
    }

    private fun recursion(n: Int, apeach: IntArray, lion: IntArray, curIdx: Int) {
        if (n == 0) {
            val diff = getPointDiff(apeach, lion)
            if (maxDiff < diff) {
                maxDiff = diff
                maxDiffArray = lion.clone()
                return
            }
            if (maxDiff > 0 && maxDiff == diff && containsMoreLowerPoints(lion)) maxDiffArray = lion.clone()
            return
        }
        for (i in curIdx until apeach.size) {
            val maxAdditionNum = min(apeach[i] + 1, n)
            lion[i] = maxAdditionNum
            recursion(n - maxAdditionNum, apeach, lion, i + 1)
            lion[i] = 0
        }
    }

    private fun getPointDiff(apeach: IntArray, lion: IntArray): Int {
        var apeachSum = 0
        var lionSum = 0
        for (i in apeach.indices) {
            when {
                apeach[i] == 0 && lion[i] == 0 -> continue
                apeach[i] >= lion[i] -> apeachSum += 10 - i
                else -> lionSum += 10 - i
            }
        }
        return lionSum - apeachSum
    }

    private fun containsMoreLowerPoints(lion: IntArray): Boolean {
        // maxDiffArray 가 더 큰게 나오기 전에 lion 이 더 큰게 나와야 한다
        for (i in lion.indices.reversed()) {
            if (lion[i] > maxDiffArray[i]) return true
            if (lion[i] < maxDiffArray[i]) return false
        }
        return false
    }
}
